### ISSUE TITLE
GradType: a subset of dtype that is differentiable, containing all float and complex dtypes

### ISSUE THREAD

**Issue by dvorst** [pytorch/pytorch#128793]
### 🚀 The feature, motivation and pitch

In the documentation, dtype is often used as a type hint where in actuality only differentiable dtypes are allowed (float and complex). Therefore the request to add GradType, a type that can have a gradient / is differentiable, which encompasses all different float and complex types. This allows type hints to more precisely indicate what is allowed.

I would like to implement this feature myself if accepted.

**Comment by mikaylagawarecki** [comment]
Do you have specific examples of functions that are type-hinted with `dtype` that would benefit from this

**Comment by dvorst** [comment]
I suppose everything for which a gradient is calculated during backprop, such as torch.nn.Linear

**Comment by ezyang** [comment]
I'm not convinced this is all that much clarity, it's usually contextually clear if something is differentiable...
